[3124] Custom Resources rewrite

damacus opened the pull request:
- Adds a new menu point for Custom Resources
- Move all custom resource docs to the custom_resource folder
- Move all the DSL pages under custom_resources/dsl so it's easier to maintain
- Add LWRP to Custom Resources guide
- Add Definition to Custom Resources guide
- Remove two shortcodes that should be full pages (resource_name & ruby_type)
- Add DSL page on: coerce, action_class & accumulators

damacus committed:
Move partials into the main glossary doc

Whilst they make editing each section easier, it's also easier to make a mistake with heading sizes.
Also the partials didn't actually work. We'll probably revet this in the future.

damacus committed:
Fix partials menu item

Orders the partials in the menu.

damacus committed:
Fix the heading for LWRP to Custom resource guide

- Fix the menu wording
- Moves it down the menu

damacus committed:
Reorder library helpers 

Move it higher in the menu as users should know about this before other topics

damacus committed:
Fix the accumulators menu item

This wasn't appearing in the menu

damacus committed:
Fix definition to custom resource menu item

This was trying to hang off the resource item rather than the custom resources section

damacus committed:
Add current_value_does_not_exist

Add stub for current_value_does_not_exist. We need a proper example here.

damacus commented on content/custom_resources/glossary.md:
````diff
@@ -0,0 +1,505 @@
+```ruby
+load_current_value do
+  current_value_does_not_exist!
+end
````
@lamont-granquist/@tas50 I'm struggling to think of a good use of this here. Any ideas?

tas50 replied:
https://github.com/chef/chef/blob/master/lib/chef/resource/windows_printer_port.rb#L93

damacus committed:
Move new_resource to the glossary

Move it to the glossary and note the new_resource is good practise.

damacus commented on content/custom_resources/glossary.md:
```diff
@@ -0,0 +1,600 @@
+
+Custom resources are designed to use core resources that are built into Chef. In some cases, it may be necessary to specify a property in the
+custom resource that is the same as a property in a core resource, for the purpose of overriding that property when used with the custom
+resource. It is still consider good practise to use `new_resource.property_name` to access the correct value.
```
Does anyone else think this section could just be "always use new_resource.property" that way you don't get collisions?

tas50 replied:
You always have to do this now. We made that a requirement.

damacus replied:
Cool thanks. I couldn't figure out if that was the case or I just *did* this out of habbit.

damacus commented:
Does anyone have a suggestion about how to do a directory tree without he linter complaining? 
The characters it doesn't like right now are all pasted from the terminal and valid.

damacus committed:
Add a note to the identity block

The identity block is no longer used and should be removed.
Keeping around so people who search for it know to remove it.

Signed-off-by: Dan Webb <dan.webb@damacus.io>

damacus committed:
Update About page title

Signed-off-by: Dan Webb <dan.webb@damacus.io>

damacus committed:
Remove DSL method files that are now empty

We've move these files into the glossary for now

Signed-off-by: Dan Webb <dan.webb@damacus.io>

damacus committed:
Move the validation_parameter back to shortcodes

Signed-off-by: Dan Webb <dan.webb@damacus.io>

marcparadise commented on content/custom_resources/definitions_to_custom_resources.md:
```diff
@@ -33,29 +30,28 @@ Though a definition looks like a resource, and at first glance seems
+- Are not true resources
+- Are processed when resource collection is compiled, not when a node
+  is converged
+- Don't support common resource properties, such as `notifies`,
```
Some have been changed to "do not" and this one is still "don't".

tas50 replied:
```suggestion
- Don't support common resource properties, such as `notifies`, `compile_time`, 
```

marcparadise commented on content/custom_resources/definitions_to_custom_resources.md:
```diff
@@ -33,29 +30,28 @@ Though a definition looks like a resource, and at first glance seems
-    provide access to parameters and their values
--   The body of the definition
+- A resource name
+- Zero or more arguments that define parameters their default values;
```
parameters ^and their default values

marcparadise commented on content/custom_resources/lwrp_to_custom_resources.md:
```diff
@@ -0,0 +1,123 @@
+
+## Convert files to Custom Resources Layout
+
+LWRPs are a pair of library files one provider, one resource.
```
Suggest: LWRPs consist of two library files -- a resource and a provider for that resource.

tas50 commented on content/custom_resources/accumulators.md:
```diff
@@ -0,0 +1,46 @@
+
+An accumulator is a programming pattern to gather multiple values together.
+
+In the context of Custom Resources, this means we collect a set of inputs, usually a set of properties from multiple Custom Resources, and then apply the final collection to a resource.
```
We need a different way to explain this. Probably with an example showing 2 resources being called in a recipe

damacus replied:
OK, I think this now covers that.

tas50 commented on content/custom_resources/definitions_to_custom_resources.md:
```diff
@@ -1,25 +1,22 @@
-ChefSpec unit testing.
+## Migrating from Definitions to Custom Resources
+
+It is no longer recommended to write resources in the Definitions format.
```
```suggestion
The definitions feature in Chef Infra has been deprecated and will be removed in a future release. Please migrate existing definitions to Custom Resources.
```

tas50 commented on content/custom_resources/definitions_to_custom_resources.md:
```diff
@@ -33,29 +30,28 @@ Though a definition looks like a resource, and at first glance seems
+- Do not support `why-run` mode
+- Cannot report to Chef Automate
+- Cannot be tested with ChefSpec
+- Some Definition paramters have known bugs, and will not be fixed
```
```suggestion
- Some Definition parameters have known bugs, and will not be fixed
```

marcparadise commented on content/custom_resources/lwrp_to_custom_resources.md:
```diff
@@ -0,0 +1,123 @@
+
+## Drop LWRP classes
+
+LWRPs required telling Chef, that they were not regular library files, but Providers and Resources. This is now taken care of for us.
```
Recommend similar to: LWRPs used classes to separate Provider and Resource behaviors, but custom resources do not need this distinction. This means that we remove the class definitions in their entirety, as shown in the following example:

marcparadise commented on content/custom_resources/lwrp_to_custom_resources.md:
```diff
@@ -0,0 +1,123 @@
+
+It is best practise to use properties to change the behaviour of resources.
+
+In the example example we used an attribute to change the `installer_url`.
```
In the *previous example we used...

marcparadise commented on content/custom_resources/lwrp_to_custom_resources.md:
````diff
@@ -0,0 +1,123 @@
+end
+```
+
+Replace the above files with a singular resource:
````
*with a single resource file, `rvm_ruby.rb`:

tas50 commented on content/custom_resources/glossary.md:
```diff
@@ -0,0 +1,607 @@
+
+## new_resource.property
+
+Custom resources are designed to use core resources that are built into Chef. In some cases, it may be necessary to specify a property in the
```
```suggestion
Custom resources are designed to use core resources that are built into Chef Infra. In some cases, it may be necessary to specify a property in the
```

tas50 commented on content/custom_resources/glossary.md:
```diff
@@ -0,0 +1,607 @@
+
+## sensitive
+
+Introduced: Chef Infra Client 12.14
```
```suggestion
```

tas50 commented on content/custom_resources/glossary.md:
````diff
@@ -0,0 +1,607 @@
+property :cloud_password, String, desired_state: false
+```
+
+## identity
````
Just nuke this whole thing

tas50 commented on content/custom_resources/glossary.md:
```diff
@@ -0,0 +1,607 @@
+
+## run_context
+
+Chef loads and tracks the current run in the run context object.
```
We need a better description of this

damacus replied:
Yeah, I could really do with jumping on a call or chat to discuss this one. I've implemented in both HAProxy and Samba.I know the how, but not the *why*. 

For example I know we need to force things into the `:root` run_context for[ multiple resources to find each other. ](https://github.com/sous-chefs/samba/blob/c1f14f2f81d7f78dafcef44e1abfc6024dca4b1f/resources/server.rb#L147-L151), but I don't know how the run_context object is structured.

marcparadise commented on content/custom_resources/lwrp_to_custom_resources.md:
```diff
@@ -0,0 +1,123 @@
+
+## Remove Attributes
+
+It is best practise to use properties to change the behaviour of resources.
```
Something like: Properties changeable values that a resource needs to perform its work. Properties are declared as part of custom resource, and much like built-in resources support validation and default values.

marcparadise commented on content/custom_resources/lwrp_to_custom_resources.md:
```diff
@@ -0,0 +1,123 @@
+
+In the example example we used an attribute to change the `installer_url`.
+
+Instead, we should use a property that we can perform checks on. In this case, we can make sure we only accept a String.
```
Consistent with best practices, we'll declare `installer_url` as a property of our custom resource.  In this example we also ensure that it will accept only a String value, and provide a reasonable default;

marcparadise commented on content/custom_resources/partials.md:
```diff
@@ -0,0 +1,40 @@
+
+[InfoQ article](https://www.infoq.com/news/2020/05/chef-infra-16/)
+
+Resoruce partials are a way of DRYing up resource code.
```
*Resource

marcparadise replied:
"DRYing" is a bit jargony for a general audience. Maybe something like "Resource partials are a way of allowing resources to share common code.  We recommend using partials any time you need to share code across three or more resources.

damacus replied:
Sounds good!

marcparadise commented on content/custom_resources/partials.md:
```diff
@@ -0,0 +1,40 @@
+
+Resoruce partials are a way of DRYing up resource code.
+
+If you have three resources all which require the proprties, `user` and `group` to be set. You can write the property definition in one file and include it in the three resources.
```
Suggest: For example let's say we have three resources that all require the properties `user` and `group`. Instead of writing those as `property` definitions in each of the three custom resources, we can write them in one file:

marcparadise commented on content/custom_resources/partials.md:
````diff
@@ -0,0 +1,40 @@
+        String,
+        default: 'haproxy'
+```
+
````
Suggest inserting: Then include them in each of the three resources with the `use` directive:

marcparadise commented on content/custom_resources/accumulators.md:
```diff
@@ -0,0 +1,46 @@
+- the Chef template resource
+- resource `[run_context](dsl/run_context.md)`
+
+You will be able to use the accumulator pattern and be aware of it's draw backs.
```
*its

damacus replied:
Thanks!

IanMadd commented:
Rename `custom_resources/about.md` to `custom_resources/_index.md`, otherwise there's a blank page at https://deploy-preview-3185--chef-web-docs.netlify.app/custom_resources/

marcparadise commented on content/custom_resources/accumulators.md:
```diff
@@ -0,0 +1,46 @@
+
+## Scope
+
+We will cover
```
Looks like this is incomplete? "Drawbacks" too.

damacus replied:
Yup. I was going to structure a lot of the sections like this but I don't think it's needed any more.

IanMadd commented on .mdlrc:
```diff
@@ -1 +1 @@
-rules "~MD013", "~MD032"
+rules "~MD013", "~MD032", "~MD041"
```
We don't need this change since we never have a heading at the top of our pages.

damacus replied:
This is why it's needed, unless I'm missing something here?

IanMadd replied:
The 041 throws an error if we're missing first line H1 headings, but we don't have them or use them

damacus replied:
Yeah, this should stop it from throwing an error. ~MD041 is ignore this rule.

IanMadd replied:
041 is the rule itself. https://github.com/DavidAnson/markdownlint/blob/main/doc/Rules.md#md041---first-line-in-a-file-should-be-a-top-level-heading

kagarmoe replied:
There's a more complete `markdowlint.yaml` in the repo.

IanMadd commented:
> Does anyone have a suggestion about how to do a directory tree without he linter complaining?
> The characters it doesn't like right now are all pasted from the terminal and valid.

We've done something like:

```
some_dir
       |- some_file.txt
       |- some_other_dir
           |- another_file.txt
```

damacus committed:
Update content/custom_resources/definitions_to_custom_resources.md

Co-authored-by: Tim Smith <tsmith@chef.io>

damacus committed:
Update content/custom_resources/definitions_to_custom_resources.md

Co-authored-by: Tim Smith <tsmith@chef.io>

damacus committed:
Apply suggestions from code review

Co-authored-by: Tim Smith <tsmith@chef.io>

xorima commented on content/custom_resources/about.md:
```diff
@@ -1,66 +1,56 @@
-in a custom resource, like "name". For example,
-`property :property_name` in the following invalid syntax:
-`property :name, String, default: 'thename'`.
+The first action listed is the default action.
```
I feel like this should have a heading around default actions to make it easier to see etc.

damacus replied:
This is covered in the glossary ##default_action

xorima commented on content/custom_resources/about.md:
```diff
@@ -86,20 +76,24 @@ action :delete do
-    not the default action)
+- `homepage` is a property that sets the default HTML for the
+  `index.html` file with a default value of `'<h1>Hello world!</h1>'`
+- the `action` block uses the built-in collection of resources to tell
```
```suggestion
- The `action` block uses the built-in collection of resources to tell
```

damacus replied:
These aren't full sentences, so don't need capitalising in this case.

xorima commented on content/custom_resources/about.md:
```diff
@@ -242,32 +236,9 @@ matters!
-package 'httpd' do
-  action :install
-end
+package 'httpd' # Ommiting the action uses the default action and properties on the resource
```
I feel like it would be better to write out the full implementation then show the short after:

```
# These are equivalent 
package 'httpd' do
  action :install
end

package 'httpd' # Ommiting the action uses the default action and properties on the resource
```

damacus replied:
👍🏼 Agreed

damacus committed:
Update accumulators

Signed-off-by: Dan Webb <dan.webb@damacus.io>

damacus committed:
Update the LWRP descriptions

Signed-off-by: Dan Webb <dan.webb@damacus.io>
Co-authored-by: Marc A. Paradise <marc.paradise@gmail.com>
Signed-off-by: Dan Webb <dan.webb@damacus.io>

damacus committed:
Update package install example

Signed-off-by: Dan Webb <dan.webb@damacus.io>

damacus committed:
Fix linting

Signed-off-by: Dan Webb <dan.webb@damacus.io>

xorima commented:
Should there also be a page on converting hwrp into custom resources?

damacus commented:
> Should there also be a page on converting hwrp into custom resources?

I don't think we really want to mention yet another way that wasn't fully supported. Unless you know of a community cookbook that still has HWRPs kicking around?

tas50 commented:
HWRPs are very common at customers still

kagarmoe commented:
Rebase to pick up the new markdownlint config. You'll need to install markdownlint-cli2 https://github.com/chef/chef-web-docs/blob/master/content/style_tools.md#markdownlint

damacus commented:
> HWRPs are very common at customers still

OK cool I think this is one for another PR then this is too bad as is.

xorima commented:
Hey,

Any idea when this will get merged in as it would be great for a few discussions I have had to point them to this rather than have to explain things which are not documented well today

I've been looking forward to this being merged for a couple months now

cc: @damacus @IanMadd @kagarmoe @bennyvasquez

damacus commented:
Paging @chef/docs-team. Let us know when you've got time to look at this one 👍🏼

tas50 commented:
@damacus can you rebase this?

tas50 commented on content/custom_resources/_index.md:
```diff
@@ -1,66 +1,56 @@
 
-action :action_name do
+action :action_a do
  # a mix of built-in Chef resources and Ruby
```
```suggestion
 # a mix of built-in Chef Infra resources and Ruby
```

tas50 commented on content/custom_resources/_index.md:
```diff
@@ -1,66 +1,56 @@
 
-action :another_action_name do
+action :action_b do
  # a mix of built-in Chef resources and Ruby
```
```suggestion
 # a mix of built-in Chef Infra resources and Ruby
```

tas50 commented on content/custom_resources/_index.md:
```diff
@@ -1,66 +1,56 @@
-### Example
+### Example Resource
 
 This example `site` utilizes Chef's built-in `file`, `service` and
```
```suggestion
This example `site` utilizes Chef Infra's built-in `file`, `service` and
```

tas50 commented on content/custom_resources/accumulators.md:
```diff
@@ -0,0 +1,106 @@
+
+This is an advanced topic. You should have already written a Custom Resource and be familiar with:
+
+- the Chef template resource
```
```suggestion
- the Chef Infra `template` resource
```

tas50 commented on content/custom_resources/accumulators.md:
```diff
@@ -0,0 +1,106 @@
+
+- delcare a resource
+- edit the resource
+- put both resources in the :root run_context
```
```suggestion
- put both resources in the :root `run_context`
```

tas50 commented on content/custom_resources/glossary.md:
```diff
@@ -0,0 +1,578 @@
+
+### Example
+
+You have a template that requires `'yes'` or `'no'` written as a String, but you would like the user to use `true` or `false` for convenience. To allow both the `:add` and `:remove` actions to have access to this method, place the method in the `action_class` block.
```
```suggestion
You have a template that requires `'yes'` or `'no'` written as a `String`, but you would like the user to use `true` or `false` for convenience. To allow both the `:add` and `:remove` actions to have access to this method, place the method in the `action_class` block.
```

tas50 commented on content/custom_resources/glossary.md:
```diff
@@ -0,0 +1,578 @@
+end
+
+action :create do
+
```
```suggestion
```

tas50 commented on content/custom_resources/glossary.md:
```diff
@@ -0,0 +1,578 @@
+- `user new_resource.user`
+- `sensitive new_resource.sensitive`
+
+correctly use the properties of the **execute** resource and not the identically-named override properties of the custom resource.
```
```suggestion
Correctly use the properties of the **execute** resource and not the identically-named override properties of the custom resource.
```

tas50 commented on content/custom_resources/glossary.md:
```diff
@@ -0,0 +1,578 @@
+
+## sensitive
+
+
```
```suggestion
```

tas50 commented on content/custom_resources/glossary.md:
```diff
@@ -0,0 +1,578 @@
+
+A property can be marked sensitive by specifying `sensitive: true` on
+the property. This prevents the contents of the property from being
+exported to data collection and sent to an Automate server.
```
Or shown in the logs of Chef Infra Client run, which is usually what people care about

tas50 commented on content/custom_resources/glossary.md:
```diff
@@ -0,0 +1,578 @@
+
+## property_is_set?
+
+Use the `property_is_set?` method to check if the value for a property is set. The syntax is:
```
We should clarify what we mean by set. The user set it in the cookbook

tas50 commented on content/custom_resources/glossary.md:
```diff
@@ -0,0 +1,578 @@
+
+### Introduced
+
+Chef Infra Client <!-- TODO:  -->
```
This has been around forever. We just require it now

tas50 commented on content/custom_resources/glossary.md:
```diff
@@ -0,0 +1,578 @@
+
+## coerce
+
+`coerce` is used to transform user input into a canonical form. The
```
"canonical form" is a really confusing way of describing what we're doing here

damacus replied:
Indeed. I'm not sure what that came from (definitely not a term I use regularly 😆 )

tas50 commented:
This is certainly an improvement, but I think there's still a lot we could do here to really clarify the mess of methods and stuff that's all shoved in the glossary. A lot of it is property related and it seems like we can focus a portion of the docs around the properties. They're the API the user interacts with and we have a lot of functionality to validate the data the user provides or transform it.

marcparadise commented on content/custom_resources/accumulators.md:
```diff
@@ -0,0 +1,106 @@
+
+If the software you are automating has a single configuration file, but you wish to split out each configuration section into its own resource this can be done with the accumulator pattern.
+
+For example, Samba uses a single file for configuration. You may wish to share multiple directories within the same configuration file. This is not be possible without multiple `samba_share` resources.
```
"This is not be possible"

damacus committed:
[3124] Update Custom Resource Documentation

- Adds a new menu point for Custom Resources
- Move all custom resource docs to the custom_resource folder
- Move all the DSL pages under custom_resources/dsl so it's easier to maintain
- Add LWRP to Custom Resources guide
- Add Definition to Custom Resources guide
- Remove two shortcodes that should be full pages (resource_name & ruby_type)
- Add DSL page on: coerce, action_class & accumulators

Signed-off-by: Dan Webb <dan.webb@damacus.io>

damacus commented:
I think there's a point where we need to iterate on these docs rather than try to make this perfect. Whilst this is an improvement this is a first step in the right direction. A lot of the wording that we can agree is incorrect, wasn't written by myself and fixing it in this Pull Request seems like it's extending the scope too much.

damacus committed:
Apply fixes from review

Signed-off-by: Dan Webb <dan.webb@damacus.io>

damacus committed:
Remove dsl_custom_resource page

Commit:
Use the markdownlint.yaml

Signed-off-by: kagarmoe <kgarmoe@chef.io>

Commit:
Light edits

Signed-off-by: kagarmoe <kgarmoe@chef.io>

Commit:
Rebased and updated nav

Signed-off-by: kagarmoe <kgarmoe@chef.io>

Commit:
copy

Signed-off-by: kagarmoe <kgarmoe@chef.io>

Commit:
Documentation edits

Resetting because of rebase failure.

Signed-off-by: kagarmoe <kgarmoe@chef.io>

xorima commented:
Anyone know when these will be merged? It's been 6 months and it would be great to get this in?

kagarmoe commented:
@Xorima We're working on it but can't commit to a publication date. There's still quite a bit of content revision involved.

xorima commented:
@kagarmoe is the content that needs revising fundamentally wrong today or can we get this merged and make small edits to it after? The value of this PR for the practitioners of Chef is huge and the longer it is until this is merged the more likely people are following incorrect practices.

I get that having perfect content is important, and this might not be perfect but merging this in and then making small changes seems more sensible imo, or asking @damacus to split thing into lots of smaller PRs and working through one at a time. It's been half a year and the fact that this still had no ETA on merging is beyond frustrating.

What action needs taking to get this either split up or merged promptly?

tas50 commented on content/chef_repo.md:
```diff
@@ -137,30 +86,23 @@ and for vim, do something like:
-the version history of each file. This is because Chef Infra Client will
-never use it while configuring nodes, plus the amount of data in an
-upload that includes top-level Subversion data could be significant.
+If using Subversion as the version source control application, it is important not to upload certain files that Subversion uses to maintain the version history of each file. This is because Chef Infra Client will never use it while configuring nodes, plus the amount of data in an upload that includes top-level Subversion data could be significant.
```
No one in 2021 is using subversion. This should go away

kagarmoe replied:
I'm always happy to curate the content!

tas50 commented on content/definitions_to_custom_resources.md:
```diff
@@ -1,20 +1,24 @@
-This topic covers migrating existing definitions to custom resources. Custom resources are integral to the modern Chef Infra workflow. While definitions are not yet deprecated, we *strongly* advise migrating to custom resources immediately. You'll benefit from the many features of Chef Infra such as notifications, reporting, why-run mode, and ChefSpec unit testing.
+## Migrating from Definitions to Custom Resources
+
+The definitions feature in Chef Infra has been deprecated and will be removed in a future release. Please migrate existing definitions to Custom Resources.
```
We have not yet committed to removing the feature. We're just strongly advocating for folks getting off them.

tas50 commented on content/lwrp_to_custom_resources.md:
```diff
@@ -0,0 +1,123 @@
+
+## Convert files to Custom Resources Layout
+
+LWRPs consist of two library files: a resource and a provider for that resource.
```
If there are two library files it's not an LWRP. It's an HWRP. LWRPs have a file in /resources and a file in /providers.

tas50 replied:
This whole page is really about HWRPs. It should just be a guide to converting from HWRPs

kagarmoe replied:
I'm opening a 'modernization' pr in #3586. We'll fix it there.

lamont-granquist commented on content/accumulators.md:
```diff
@@ -0,0 +1,106 @@
+      action :nothing
+      delayed_action :create
+    end
+end
```
you're missing an `end` here and on the action down below

kagarmoe commented on content/accumulators.md:
```diff
@@ -0,0 +1,106 @@
+This is an advanced topic. You should have already written a Custom Resource and be familiar with:
+
+- the built-in Chef Infra `template` resource
+- resource `[run_context](dsl/run_context.md)`
```
Should this be `with_run_context` from the [cookbook execution](https://docs.chef.io/infra_language/cookbook_execution/#with_run_context) document?

```suggestion
- Cookbook execution [`with_run_context`]({{< relref "infra_language/cookbook_execution/#with_run_context" >}})
```

lamont-granquist commented on content/accumulators.md:
```diff
@@ -0,0 +1,106 @@
+      variables(samba_options: new_resource.options)
+      action :nothing
+      delayed_action :create
+    end
```
I sort of feel like both of these resources should call a method that uses edit_resource to assert the template resource is there and not declare it twice.  That way if you call `samba_share` before `samba_server` it doesn't create two different template resources (or you can call samba_server twice or other situations), but that may distract from the example.  At the same time it feels like if we're going to have an example it shouldn't be a pattern with edge conditions to it, but then I'd extract the method out into a resource partial and that starts to get irrelevant to the accumulator pattern...

kagarmoe replied:
I think we need to break updating the accumulator documentation into a separate piece of work. Happily, there's already an issue for it: #3123

lamont-granquist commented on content/accumulators.md:
````diff
@@ -0,0 +1,106 @@
+_Note_: We use the `||=` Ruby method to add a new Hash if one does not exist already.
+
+```ruby
+# samba_server
````
It actually took me a second to realize these were actions in two different resources.  And the text above that doesn't address it, this is the only mention of `samba_server` so I think it might not be entirely clear why its here.

kagarmoe replied:
I tried to clarify the two different resources (check the netlify preview) but I'll admit that I don't understand how this all hangs together.

lamont-granquist commented on content/helpers.md:
```diff
@@ -0,0 +1,42 @@
+    weight = 60
++++
+
+Helper classes enable users to share code between Custom Resources. Rspec method test are possible when code is abstracted into plain Ruby files.
```
Might be good to push people towards using resource partials here rather than helpers.

tas50 replied:
Helpers are pretty terrible. Anything we could do here would be helpful.

kagarmoe replied:
We need to scope a "Using helpers and partials" issue.

kagarmoe commented:
@Xorima I think we're nearing the end of the rewrite! If you have the time, would you also read it through and add your comments?

Here's the link to the preview. Let me know if you can't see it.
https://deploy-preview-3185--chef-web-docs.netlify.app/custom_resources/

xorima reviewed (approved):
@kagarmoe  all happy from my end, anything I found had already been commented on above

Commit:
Merge branch 'main' into 3124/custom-resources-rewrite

Commit:
Markdownlint

Signed-off-by: kagarmoe <kgarmoe@chef.io>

damacus commented on themes/docs-new/layouts/shortcodes/identity.md:
```diff
@@ -1,3 +1,6 @@
+
+TODO: this sentence doesn't make any sense yet
+
```
```suggestion
```

damacus committed:
Update themes/docs-new/layouts/shortcodes/identity.md

Commit:
Merge branch 'main' into 3124/custom-resources-rewrite

kagarmoe reviewed (approved):
Ok to merge

gene1wood commented on content/custom_resources.md:
```diff
@@ -86,471 +79,28 @@ action :delete do
+### Syntax
 
-### coerce
+Call a resource in a recipe by its `resource_name`. For example:
```
It looks like the example that this line references wasn't included
